fedimint-client/src/module/init/recovery.rs

```diff
@@ -247,24 +247,32 @@ where
             api: DynGlobalApi,
             decoders: ModuleDecoderRegistry,
             epoch_range: ops::Range<u64>,
-        ) -> impl futures::Stream<Item = (u64, SessionOutcome)> + 'a {
+        ) -> impl futures::Stream<Item = (u64, Vec<AcceptedItem>)> + 'a {
             // How many request for blocks to run in parallel (streaming).
             const PARALLISM_LEVEL: usize = 8;
 
-            futures::stream::iter(epoch_range)
-                .map(move |block_idx| {
+            futures::stream::iter(epoch_range.clone())
+                .map(move |session_idx| {
                     let api = api.clone();
                     let decoders = decoders.clone();
                     Box::pin(async move {
-                        info!(block_idx, "Fetching epoch");
+                        info!(session_idx, "Fetching epoch");
 
                         let mut retry_sleep = Duration::from_millis(10);
                         let block = loop {
-                            info!(target: LOG_CLIENT_RECOVERY, block_idx, "Awaiting signed block");
-                            match api.await_block(block_idx, &decoders).await {
+                            info!(target: LOG_CLIENT_RECOVERY, session_idx, "Awaiting signed block");
+
+                            let items_res = api.get_session_status(session_idx, &decoders).await.map(|s| match s {
+                                    SessionStatus::Initial => panic!("Federation missing a block that existed when we started recovery"),
```

Comment:
Is this possibly racy @joschisan? If we get the current session height and we are in between session the value returned could reasonably be `Initial` afaik? Or is the session switch atomic and initial will only be returned for strictly future sessions?

I don't think this matters too much because the session switch time is likely on the order of milli seconds and recovery will take much longer, just want to validate my mental model.

Reply:
One benefit of https://github.com/fedimint/fedimint/pull/4159 is that session count and session status come from the same source in the database, so there should be any room for races anymore (thought I'm not 100% of the db isolation level thingies).